ros_controllers/src/diff_drive_controller.cpp

```diff
@@ -14,12 +14,272 @@
+
+DiffDriveController::DiffDriveController() : controller_interface::ControllerInterface() {}
+
+DiffDriveController::DiffDriveController(std::vector<std::string> left_wheel_names,
+                                         std::vector<std::string> right_wheel_names,
+                                         std::vector<std::string> write_op_names) :
+   controller_interface::ControllerInterface(),
+   left_wheel_names_(std::move(left_wheel_names)),
+   right_wheel_names_(std::move(right_wheel_names)),
+   write_op_names_(std::move(write_op_names))
+{}
+
+controller_interface::controller_interface_ret_t
+DiffDriveController::init(std::weak_ptr<hardware_interface::RobotHardware> robot_hardware,
+                          const std::string& controller_name)
+{
+   // initialize lifecycle node
+   auto ret = ControllerInterface::init(robot_hardware, controller_name);
+   if (ret != CONTROLLER_INTERFACE_RET_SUCCESS) {
+      return ret;
+   }
+
+   // with the lifecycle node being initialized, we can declare parameters
+   lifecycle_node_->declare_parameter<std::vector<std::string>>("left_wheel_names", left_wheel_names_);
+   lifecycle_node_->declare_parameter<std::vector<std::string>>("right_wheel_names", right_wheel_names_);
+   lifecycle_node_->declare_parameter<double>("wheel_separation", wheel_params_.separation);
+   lifecycle_node_->declare_parameter<int>("wheels_per_side", wheel_params_.wheels_per_side);
+   lifecycle_node_->declare_parameter<double>("wheel_radius", wheel_params_.radius);
+   lifecycle_node_->declare_parameter<double>("wheel_separation_multiplier", wheel_params_.separation_multiplier);
+   lifecycle_node_->declare_parameter<double>("left_wheel_radius_multiplier", wheel_params_.left_radius_multiplier);
+   lifecycle_node_->declare_parameter<double>("right_wheel_radius_multiplier", wheel_params_.right_radius_multiplier);
+   lifecycle_node_->declare_parameter<std::string>("odom_frame_id", odom_frame_id_);
+   lifecycle_node_->declare_parameter<std::string>("base_frame_id", base_frame_id_);
+   lifecycle_node_->declare_parameter<std::vector<std::string>>("write_op_modes", write_op_names_);
+
+   return CONTROLLER_INTERFACE_RET_SUCCESS;
+}
+
+controller_interface::controller_interface_ret_t DiffDriveController::update()
+{
+   auto logger = lifecycle_node_->get_logger();
+   if (lifecycle_node_->get_current_state().id() == State::PRIMARY_STATE_INACTIVE) {
+      if (!is_halted) {
+         halt();
+         is_halted = true;
+      }
+      return CONTROLLER_INTERFACE_RET_SUCCESS;
+   }
+
+   // Apply (possibly new) multipliers:
+   const auto wheels = wheel_params_;
+
+   const double wheel_separation = wheels.separation_multiplier * wheels.separation;
+   const double left_wheel_radius = wheels.left_radius_multiplier * wheels.radius;
+   const double right_wheel_radius = wheels.right_radius_multiplier * wheels.radius;
+
+   // Compute wheels velocities:
+   const auto& current_command = *velocity_msg_ptr_;
```

Comment:
```suggestion
   const auto& current_command = velocity_msg_ptr_.get();
```

Also verify the velocity is not null.

Reply:
Yeah good call. 

I was trying to dereference the pointer right there instead of working with the raw pointer, so I can use the dot operator and call a clean variable rather than the straight pointer.